File: app/src/main/java/com/kieran/winnipegbusbackend/UpcomingStops/HttpUpcomingStopsManager.java
package com.kieran.winnipegbusbackend.UpcomingStops;

import com.kieran.winnipegbusbackend.LoadResult;
import com.kieran.winnipegbusbackend.RouteKey;
import com.kieran.winnipegbusbackend.SearchQuery;
import com.kieran.winnipegbusbackend.TransitApiManager;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class HttpUpcomingStopsManager implements UpcomingStopsManager {
    @Override
    public void GetUpcomingStopsAsync(RouteKey key, int stopOnRoute, final OnUpcomingStopsFoundListener listener) {
        SearchQuery query = TransitApiManager.generateSearchQuery(key);
        TransitApiManager.getJsonAsync(query.getQueryUrl(), new TransitApiManager.OnJsonLoadResultReceiveListener() {
            @Override
            public void OnReceive(LoadResult<JSONObject> result) {
                LoadResult<ArrayList<Integer>> loadResult = new LoadResult<>(null, null);

                if (result.getResult() != null) {
                    try {
                        JSONArray stops  = (result.getResult()).getJSONArray("stops");
                        ArrayList<Integer> stopNumbers = new ArrayList<>();

                        for (int i = 0; i < stops.length(); i++) {
                            stopNumbers.add(stops.getJSONObject(i).getInt("number"));
                        }

                        loadResult.setResult(stopNumbers);
                    } catch (JSONException ex) {
                        loadResult.setException(ex);
                    }
                }

                listener.OnUpcomingStopsFound(loadResult);
            }
        });
    }
}

File: app/src/main/java/com/kieran/winnipegbusbackend/UpcomingStops/HttpUpcomingStopsManager.kt
package com.kieran.winnipegbusbackend.UpcomingStops

import com.kieran.winnipegbusbackend.LoadResult
import com.kieran.winnipegbusbackend.RouteKey
import com.kieran.winnipegbusbackend.winnipegtransit.TransitApiManager

import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

class HttpUpcomingStopsManager : UpcomingStopsManager {
    override fun GetUpcomingStopsAsync(key: RouteKey, stopOnRoute: Int, listener: UpcomingStopsManager.OnUpcomingStopsFoundListener) {
        val query = TransitApiManager.generateSearchQuery(key)
        TransitApiManager.getJsonAsync(query.queryUrl, object : TransitApiManager.OnJsonLoadResultReceiveListener {
            override fun onReceive(result: LoadResult<JSONObject>) {
                if (result.result != null) {
                    try {
                        val stops = result.result.getJSONArray("stops")
                        val stopNumbers = ArrayList<Int>()

                        for (i in 0 until stops.length()) {
                            stopNumbers.add(stops.getJSONObject(i).getInt("number"))
                        }

                        listener.OnUpcomingStopsFound(LoadResult(stopNumbers, null))
                    } catch (ex: JSONException) {
                        listener.OnUpcomingStopsFound(LoadResult<ArrayList<Int>>(null, ex))
                    }
                }
            }
        })
    }
}
